activation/e2e/checkpoint_test.go

```diff
@@ -0,0 +1,255 @@
+	require.NoError(t, err)
+	t.Cleanup(func() { assert.NoError(t, verifier.Close()) })
+
+	validator := activation.NewValidator(db, poetDb, cfg, opts.Scrypt, verifier)
+
+	epoch := layerDuration * time.Duration(layersPerEpoch)
+	poetCfg := activation.PoetConfig{
+		PhaseShift:  epoch,
+		CycleGap:    3 * epoch / 4,
+		GracePeriod: epoch / 4,
+	}
+	client := ae2e.NewTestPoetClient(1)
+	poetService := activation.NewPoetServiceWithClient(poetDb, client, poetCfg, logger)
+
+	// ensure that genesis aligns with layer timings
+	genesis := time.Now().Add(layerDuration).Round(layerDuration)
+	clock, err := timesync.NewClock(
+		timesync.WithGenesisTime(genesis),
+		timesync.WithLayerDuration(layerDuration),
+		timesync.WithTickInterval(100*time.Millisecond),
+		timesync.WithLogger(zap.NewNop()),
+	)
+	require.NoError(t, err)
+	t.Cleanup(clock.Close)
+
+	localDB := localsql.InMemory()
+	nb, err := activation.NewNIPostBuilder(
+		localDB,
+		svc,
+		logger.Named("nipostBuilder"),
+		poetCfg,
+		clock,
+		validator,
+		activation.WithPoetServices(poetService),
+	)
+	require.NoError(t, err)
+
+	atxdata := atxsdata.New()
+	atxVersions := activation.AtxVersions{0: types.AtxV2}
+	edVerifier := signing.NewEdVerifier()
+	mpub := mocks.NewMockPublisher(ctrl)
+	mFetch := smocks.NewMockFetcher(ctrl)
+	mBeacon := activation.NewMockAtxReceiver(ctrl)
+	mTortoise := smocks.NewMockTortoise(ctrl)
+
+	atxHdlr := activation.NewHandler(
+		"local",
+		cdb,
+		atxdata,
+		edVerifier,
+		clock,
+		mpub,
+		mFetch,
+		goldenATX,
+		validator,
+		mBeacon,
+		mTortoise,
+		logger,
+		activation.WithAtxVersions(atxVersions),
+	)
+
+	tab := activation.NewBuilder(
+		activation.Config{GoldenATXID: goldenATX},
+		db,
+		atxdata,
+		localDB,
+		mpub,
+		nb,
+		clock,
+		syncer,
+		logger,
+		activation.WithPoetConfig(poetCfg),
+		activation.WithValidator(validator),
+		activation.BuilderAtxVersions(atxVersions),
+	)
+	tab.Register(sig)
+
+	var atx0ID types.ATXID
+	mpub.EXPECT().Publish(gomock.Any(), pubsub.AtxProtocol, gomock.Any()).DoAndReturn(
+		func(ctx context.Context, p string, msg []byte) error {
+			var watx wire.ActivationTxV2
+			codec.MustDecode(msg, &watx)
+			atx0ID = watx.ID()
+			peer := peer.ID(p)
+			mFetch.EXPECT().RegisterPeerHashes(peer, gomock.Any())
+			mFetch.EXPECT().GetPoetProof(gomock.Any(), gomock.Any())
+			mBeacon.EXPECT().OnAtx(gomock.Any())
+			mTortoise.EXPECT().OnAtx(gomock.Any(), gomock.Any(), gomock.Any())
+			err := atxHdlr.HandleGossipAtx(ctx, peer, msg)
+			require.NoError(t, err)
+			return err
+		},
+	)
+
+	require.NoError(t, tab.BuildInitialPost(ctx, sig.NodeID()))
+	require.NoError(t, tab.PublishActivationTx(ctx, sig))
+
+	// Execute checpoint-recovery
+	// 1. Generate checkpoint
+	require.NoError(t, accounts.Update(db, &types.Account{}))
+	snapshot := clock.CurrentLayer()
+	fs := afero.NewMemMapFs()
+	dir, err := afero.TempDir(fs, "", "Generate")
+	require.NoError(t, err)
+	err = checkpoint.Generate(ctx, fs, db, dir, snapshot, 1)
```

Comment:
Did you look into replacing `afero` with `io/fs`? I think the biggest issue is that `io/fs` doesn't really provide write support. Does `afero` give us any benefit over working with the file system directly?

Reply:
I didn't and I'm not sure why we use it. Perhaps we should create a tech-debt issue to investigate and possibly remove afero.

Reply:
I think the original idea was to be able to pass a in-memory FS during tests, but instead we could also just pass a `t.TempDir()` and use `os` functions instead.